File: poc/src/engine/game.ts
import type { Card } from "./card";
import { Deck } from "./deck";
import { boardCellCount, type BoardSize, DEFAULT_BOARD_SIZE, Grid } from "./grid";
import { scoreGridComplete } from "./scoring";

export type GameStatus = "playing" | "finished";

export const POOL_SIZE_MIN = 1;
export const POOL_SIZE_MAX = 5;
export const DEFAULT_POOL_SIZE = 5;
export const POOL_SIZE_OPTIONS = [1, 2, 3, 4, 5] as const;

export interface SoloGameState {
  poolSize: number;
  boardSize: BoardSize;
  deck: Card[];
  pool: Card[];
  gridCells: (Card | null)[];
  turn: number;
  status: GameStatus;
  /** Cards available at shuffle time (sub-deck pressure). */
  deckSize: number;
}

export interface CreateGameOptions {
  boardSize?: BoardSize;
  deckSize?: number;
  rng?: () => number;
}

export function gridFromState(state: SoloGameState): Grid {
  return Grid.fromCells(state.gridCells, state.boardSize);
}

export function createInitialState(
  poolSize: number,
  options: CreateGameOptions | (() => number) = {},
): SoloGameState {
  let opts: CreateGameOptions;
  if (typeof options === "function") {
    opts = { rng: options };
  } else {
    opts = options;
  }
  const rng = opts.rng ?? Math.random;
  const boardSize = opts.boardSize ?? DEFAULT_BOARD_SIZE;
  const cells = boardCellCount(boardSize);
  if (poolSize < POOL_SIZE_MIN || poolSize > POOL_SIZE_MAX) {
    throw new Error(`pool_size must be ${POOL_SIZE_MIN}–${POOL_SIZE_MAX}`);
  }
  const deckSize = opts.deckSize ?? 52;
  if (deckSize < cells || deckSize > 52) {
    throw new Error(`deckSize must be ${cells}–52`);
  }
  const deck = new Deck();
  deck.shuffle(rng);
  const state: SoloGameState = {
    poolSize,
    boardSize,
    deck: deck.cards.slice(0, deckSize),
    pool: [],
    gridCells: Array(cells).fill(null),
    turn: 0,
    status: "playing",
    deckSize,
  };
  return refillPool(state);
}

function refillPool(state: SoloGameState): SoloGameState {
  const needed = state.poolSize - state.pool.length;
  if (needed <= 0 || state.deck.length === 0) {
    return state;
  }
  const drawCount = Math.min(needed, state.deck.length);
  return {
    ...state,
    deck: state.deck.slice(drawCount),
    pool: [...state.pool, ...state.deck.slice(0, drawCount)],
  };
}

export function legalCellKeys(state: SoloGameState): string[] {
  const grid = gridFromState(state);
  return grid.legalPositions().map(({ row, col }) => `${row},${col}`);
}

export function canDropOnCell(state: SoloGameState, row: number, col: number): boolean {
  return legalCellKeys(state).includes(`${row},${col}`);
}

export function placeFromPool(
  state: SoloGameState,
  poolIndex: number,
  row: number,
  col: number,
): SoloGameState {
  if (state.status !== "playing") {
    throw new Error("Game is finished");
  }
  const grid = gridFromState(state);
  if (grid.isFull()) {
    throw new Error("Grid is full");
  }
  if (poolIndex < 0 || poolIndex >= state.pool.length) {
    throw new Error("Invalid pool index");
  }
  if (!canDropOnCell(state, row, col)) {
    throw new Error("Illegal placement");
  }

  const pool = [...state.pool];
  const [card] = pool.splice(poolIndex, 1);
  grid.place(row, col, card);

  let next: SoloGameState = {
    ...state,
    pool,
    gridCells: grid.cloneCells(),
    turn: state.turn + 1,
  };
  if (pool.length === 0 && !grid.isFull()) {
    next = refillPool(next);
  }
  if (grid.isFull()) {
    next.status = "finished";
  }
  return next;
}

export function finalScore(state: SoloGameState) {
  const grid = gridFromState(state);
  if (!grid.isFull()) {
    throw new Error("Game not finished");
  }
  return scoreGridComplete(grid);
}

File: prototype/quintet/game.py
"""Solo game state machine."""

from __future__ import annotations

import random
from dataclasses import dataclass, field

from quintet.card import Card
from quintet.deck import Deck
from quintet.grid import Grid
from quintet.scoring import ScoringRule, DEFAULT_SCORING, score_grid

POOL_SIZE_MIN = 1
POOL_SIZE_MAX = 5


@dataclass
class SoloGame:
    pool_size: int
    deck: Deck = field(default_factory=Deck)
    pool: list[Card] = field(default_factory=list)
    grid: Grid = field(default_factory=Grid)
    scoring: ScoringRule = DEFAULT_SCORING
    turn: int = 0

    def __post_init__(self) -> None:
        if not POOL_SIZE_MIN <= self.pool_size <= POOL_SIZE_MAX:
            raise ValueError(f"pool_size must be {POOL_SIZE_MIN}–{POOL_SIZE_MAX}")

    def start(self, rng: random.Random | None = None) -> None:
        rng = rng or random.Random()
        self.deck.shuffle(rng)
        self.pool = []
        self.grid = Grid()
        self.turn = 0
        self._refill_pool()

    def _refill_pool(self) -> None:
        needed = self.pool_size - len(self.pool)
        if needed <= 0 or len(self.deck) == 0:
            return
        draw_count = min(needed, len(self.deck))
        self.pool.extend(self.deck.draw(draw_count))

    def legal_actions(self) -> list[tuple[int, int, int]]:
        """Returns list of (pool_index, row, col)."""
        positions = self.grid.legal_positions()
        actions: list[tuple[int, int, int]] = []
        for pi in range(len(self.pool)):
            for row, col in positions:
                actions.append((pi, row, col))
        return actions

    def play(self, pool_index: int, row: int, col: int) -> Card:
        if self.grid.is_full():
            raise ValueError("Grid is full")
        if pool_index < 0 or pool_index >= len(self.pool):
            raise ValueError("Invalid pool index")
        card = self.pool.pop(pool_index)
        self.grid.place(row, col, card)
        self.turn += 1
        if len(self.pool) == 0 and not self.grid.is_full():
            self._refill_pool()
        return card

    def is_over(self) -> bool:
        return self.grid.is_full()

    def final_score(self) -> dict:
        if not self.is_over():
            raise ValueError("Game not finished")
        return score_grid(self.grid, self.scoring)
